### PR TITLE
ISPN-6542 Subclasses of AbstractWriteKeyCommand don't marshal the Par…

### PR DESCRIPTION
**Pull request by rvansa**
…ams params field

* marshall params
* pass params to all write/read-write functional commands
* don't ignore some commands in CacheWriterInterceptor
* add Dummy store to functional API tests

### CODE REVIEW

**Comment by rvansa** [comment]
@galderz Please review.

**Comment by ksobolew** [comment]
FYI: there's one more command that is not handled in CacheWriterInterceptor but should be.

diff --git a/core/src/main/java/org/infinispan/interceptors/CacheWriterInterceptor.java b/core/src/main/java/org/infinispan/interceptors/CacheWriterInterceptor.java
index c9672a4..f6a0b5f 100755
--- a/core/src/main/java/org/infinispan/interceptors/CacheWriterInterceptor.java
+++ b/core/src/main/java/org/infinispan/interceptors/CacheWriterInterceptor.java
@@ -10,6 +10,7 @@
 import org.infinispan.commands.functional.ReadWriteManyCommand;
 import org.infinispan.commands.functional.ReadWriteManyEntriesCommand;
 import org.infinispan.commands.functional.WriteOnlyKeyCommand;
+import org.infinispan.commands.functional.WriteOnlyKeyValueCommand;
 import org.infinispan.commands.functional.WriteOnlyManyCommand;
 import org.infinispan.commands.functional.WriteOnlyManyEntriesCommand;
 import org.infinispan.commands.tx.CommitCommand;
@@ -230,6 +231,11 @@ public Object visitWriteOnlyKeyCommand(InvocationContext ctx, WriteOnlyKeyComman
       return visitWriteCommand(ctx, command);
    }
 
+   @Override
+   public Object visitWriteOnlyKeyValueCommand(InvocationContext ctx, WriteOnlyKeyValueCommand command) throws Throwable {
+      return visitWriteCommand(ctx, command);
+   }
+
    private <T extends DataWriteCommand & ParamsCommand> Object visitWriteCommand(InvocationContext ctx, T command) throws Throwable {
       Object retval = invokeNextInterceptor(ctx, command);
       if (!isStoreEnabled(command) || ctx.isInTxScope() || !command.isSuccessful()) return retval;

**Comment by rvansa** [comment]
@ksobolew thanks, an oversight.

File: core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package org.infinispan.commands.functional;
+
+import java.util.Set;
+
+/**
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
```
Comment (slaskawi) [inline core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java]: Some meaningful docs here could be useful.

**Comment by slaskawi** [comment]
I would only request this one small javadoc to be added. And could you please rebase it? I'm not sure if those CI failures are related or not.

Could you tell me what are the performance implication of this? I'm especially interested how much do we gain (?) with ordinal and how much do we loose with sending an object.

File: core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java (hunk 1 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java]: It's all it is, just a superinterface for the functional commands that operate on multiple keys.

**Comment by rvansa** [comment]
@slaskawi I've rebased it this morning and Github tells me there're no conflicts. The CI failures are IMO unrelated, this affects only Functional API tests.

I haven't benchmarked anything, I just thought about the most effective way to read/write the params (space-wise), given that those already have their disctinct IDs (so we don't have to register another ones for AdvancedExternalizers). So we can spare the lookup as well.
I don't think the actual effect will be visible.

File: core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java (hunk 1 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java]: @rvansa That would work as a javadoc :)

**Comment by danberindei** [comment]
@rvansa looks good to me
@galderz do you want to have a look as well?

**Comment by rvansa** [comment]
@galderz commented on some things in #4303, I'll update this (and remove the ManyDataCommand)

File: commons/src/main/java/org/infinispan/commons/api/functional/Param.java [hunk 2]
```diff
@@ -44,6 +44,29 @@
    P get();
 
    /**
+    * Should this paramater go over the wire?
+    */
+   SerializeMethod serializeMethod();
```
Comment (galderz) [inline commons/src/main/java/org/infinispan/commons/api/functional/Param.java]: This is not how we deal with an objects serialization/marshalling characteristics. Instead, if a param instance needs to be serialized, an Externalizer should be provided for it (examples in the codebase) and then the Externalizer it's either caught by annotation processing (e.g. @SerializedWith) or it's preregistered. The externalizer itself can decide how it serialize it using whichever method it wants. A param implementation, if you want something lazy, can implement Serializable and that's it. So, there's no need to add anything to the Param interface, which has been designed with the bare minimum required to do its job.

File: commons/src/main/java/org/infinispan/commons/api/functional/Param.java [hunk 3]
```diff
@@ -44,6 +44,29 @@
...
+
+   /**
+    * The way how this object should be serialized;
+    */
+   enum SerializeMethod {
```
Comment (galderz) [inline commons/src/main/java/org/infinispan/commons/api/functional/Param.java]: Also, whether the param itself should be sent over the wire or not is not down the parameter itself, but more to the command being executed. The command itself must decide whether it includes certain params inside its externalizer payload. Params themselves do not live on their own.

File: commons/src/main/java/org/infinispan/commons/api/functional/Param.java (hunk 2 above)
Comment (galderz) [inline commons/src/main/java/org/infinispan/commons/api/functional/Param.java]: Cross posting what I said in #4303. My feeling is that overall this method and enum pollute the interface itself when there's some clear ways to avoid the issue without the need to add extra methods.

Comment (galderz) [inline commons/src/main/java/org/infinispan/commons/api/functional/Param.java]: Sorry, we're already discussing this in #4303, let's continue there.

Comment (ksobolew) [inline commons/src/main/java/org/infinispan/commons/api/functional/Param.java]: I'm afraid I'm responsible for the mixup :) I'll explain what happened so other can learn from my mistakes ;) Basically I should've created separate PR for the backport of this PR to 8.2.x, and then base my "proposed fix" on that. But at first I thought they could be independent of each other so I only created one PR. Later on this proved to be untrue. I did create separate branches for both but it was too late to change the base branch of the PR and the backport shows as part of my PR.

**Comment by rvansa** [comment]
@danberindei @galderz @slaskawi @ksobolew Updated.

File: commons/src/main/java/org/infinispan/commons/api/functional/Param.java (hunk 2 above)
Comment (rvansa) [inline commons/src/main/java/org/infinispan/commons/api/functional/Param.java]: @ksobolew Don't worry about the mixup, as long as you keep those as separate commits, so you can easily rebase your changes. And thanks for the good work :)

**Comment by galderz** [comment]
@rvansa Thanks! Looking... :)

**Comment by galderz** [comment]
Looks great @rvansa, apologies for the delay getting around to reviewing this. Integrating...

**Comment by galderz** [comment]
Integrated, thanks @rvansa!

File: core/src/main/java/org/infinispan/functional/impl/Params.java [hunk 4]
```diff
@@ -108,4 +119,35 @@ public static Params from(Param<?>... ps) {
...
+
+      @Override
+      public void writeObject(ObjectOutput output, Params params) throws IOException {
+         // There's no point in sending FutureMode over wire
+         output.writeInt(((PersistenceMode) params.get(PersistenceMode.ID).get()).ordinal());
+      }
+
+      @Override
+      public Params readObject(ObjectInput input) throws IOException, ClassNotFoundException {
+         PersistenceMode persistenceMode = PersistenceMode.values()[input.readInt()];
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/functional/impl/Params.java]: `values()` should be cached, because it allocates a new array every time.

File: core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java [hunk 5]
```diff
@@ -384,6 +385,8 @@ private void loadInternalMarshallables() {
       addInternalExternalizer(new MarshallableFunctionExternalizers.LambdaWithMetasExternalizer());
       addInternalExternalizer(new MarshallableFunctionExternalizers.SetValueIfEqualsReturnBooleanExternalizer());
       addInternalExternalizer(new PersistentUUID.Externalizer());
+
+      addInternalExternalizer(Params.Externalizer.INSTANCE);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java]: If you always call the externalizer's methods directly, this registration will never be used. So you could remove the `Params.Externalizer` class altogether and make the methods static in `Params`.

File: core/src/main/java/org/infinispan/functional/impl/Params.java (hunk 4 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/functional/impl/Params.java]: Oops, haven't realized this. I'll issue fixup PR.

File: core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java (hunk 5 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/marshall/core/ExternalizerTable.java]: I've expected that under some circumstances it could get serialized, but ATM it would make more sense to omit the registration.
